bevy_lint/src/lints/suspicious/unit_in_bundle.rs

```diff
@@ -0,0 +1,415 @@
+            && let ty::TraitPredicate {
+                trait_ref,
+                // Negative trait bounds, which are unstable, allow matching all types _except_
+                // those with a specific trait. We don't want that, however, so we only match
+                // positive trait bounds.
+                polarity: PredicatePolarity::Positive,
+            } = trait_predicate
+            // Only match `T: Bundle` predicates.
+            && paths::BUNDLE.matches(cx, trait_ref.def_id)
+        {
+            let self_ty = trait_ref.self_ty();
+
+            debug_assert!(
+                matches!(self_ty.kind(), ty::TyKind::Param(_)),
+                "type from trait bound was expected to be a type parameter",
+            );
+
+            // At this point, we've confirmed the predicate is `T: Bundle`! Add it to the list to
+            // be returned. :)
+            bundle_bounded_generics.push(trait_ref.self_ty());
+        }
+    }
+
+    bundle_bounded_generics
+}
+
+/// Represents the path to an item within a nested tuple.
+///
+/// # Example
+///
+/// Each number within the [`TuplePath`] represents an index into the tuple. An empty path
+/// represents the root tuple, while a path of `TuplePath([0])` represents the first item within
+/// that tuple.
+///
+/// ```ignore
+/// // TuplePath([])
+/// (
+///     // TuplePath([0])
+///     Name::new("Foo"),
+///     // TuplePath([1])
+///     (
+///         // TuplePath([1, 0])
+///         (),
+///         // TuplePath([1, 1])
+///         Transform::default(),
+///         // TuplePath([1, 2])
+///         Visibility::Hidden,
+///     ),
+///     // TuplePath([2])
+///     (),
+/// )
+/// ```
+#[derive(Clone, Debug)]
+#[repr(transparent)]
+struct TuplePath(Vec<usize>);
+
+impl TuplePath {
+    /// Creates an empty [`TuplePath`].
+    fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    /// Pushes an index to the end of the path.
+    fn push(&mut self, i: usize) {
+        self.0.push(i);
+    }
+
+    /// Pops the last index in the path.
+    fn pop(&mut self) -> Option<usize> {
+        self.0.pop()
+    }
+
+    /// Finds the [`Expr`] of the item represented by this path given the root tuple.
+    ///
+    /// In the event the path is invalid in some way (such as if an expected tuple is not found),
+    /// this will return the expression closest to the target.
+    fn into_expr<'tcx>(self, root_tuple: &'tcx Expr<'tcx>) -> &'tcx Expr<'tcx> {
+        let mut tuple = root_tuple;
+
+        for i in self.0 {
+            let ExprKind::Tup(items) = tuple.kind else {
+                // If the path is invalid in some way, return the expression nearest to the target.
+                // This is usually the case when the bundle is created outside of
+                // `Commands::spawn()`, such as with `commands.spawn(my_helper())` instead of the
+                // expected `commands.spawn((Foo, Bar, ()))`.
+                return tuple;
```

Comment:
This bit can be improved, as detailed by #248, but I chose to leave it unchanged for this PR.